### ISSUE TITLE
Is clientCredentials flow supported?

### ISSUE THREAD

**Issue by markpritchett** [domaindrivendev/Swashbuckle.AspNetCore#628]
I'm trying to get `clientCredentials` flow working using one of the sample websites in this repo.

I've tried to modify `Swashbuckle.AspNetCore/test/WebSites/OAuth2Integration/Startup.cs` to use the `clientCredentials` flow type.

Given that `OAuthClientId` and `OAuthClientSecret` have been set in the `Configure` method, I hoped that I could alter the `ConfigureServices` to use the `clientCredentials` flow type and that would allow the  Swagger UI to _automagically_ retrieve the authentication token.

``` CSharp
public void Configure(IApplicationBuilder app, IHostingEnvironment env, ILoggerFactory loggerFactory)
{
    // ...
    app.Map("/resource-server", resourceServer =>
    {
        // ... 
        resourceServer.UseSwaggerUI(c =>
        {
            c.SwaggerEndpoint("/resource-server/swagger/v1/swagger.json", "My API V1");

            // Additional OAuth settings (See https://github.com/swagger-api/swagger-ui/blob/v3.10.0/docs/usage/oauth2.md)
            c.OAuthClientId("test-id");
            c.OAuthClientSecret("test-secret");
            c.OAuthRealm("test-realm");
            c.OAuthAppName("test-app");
            c.OAuthScopeSeparator(" ");
            c.OAuthAdditionalQueryStringParams(new { foo = "bar" });
            c.OAuthUseBasicAuthenticationWithAccessCodeGrant();
        });
    });
}
```

I've tried changing part of the `ConfigureServices` method to this:

``` CSharp
c.AddSecurityDefinition("Bearer", new ApiKeyScheme
{
    Description = "JWT Authorization header using the Bearer scheme. Example: \"Authorization: Bearer {token}\"",
    Name = "Authorization",
    In = "header",
    Type = "apiKey"
});
```

However, this just gives me a prompt for the bearer token, like so:

![image](https://user-images.githubusercontent.com/1267889/36868251-b1280d96-1d8f-11e8-9491-3c370ae0ea5e.png)

Looking at the [Open API Specification - Security Scheme Object Example](https://swagger.io/specification/#securitySchemeObject) it has the following JWT Bearer Sample:

``` JSON
{
  "type": "http",
  "scheme": "bearer",
  "bearerFormat": "JWT",
}
```

I couldn't see any similar options in Swashbuckle to configure the security definition.

It would be great if you could confirm whether what I am asking for is supported and if it is, provide an example.

**Comment by Lance0724** [comment]
JWT is not work in version 2.2
"Authorization: Bearer xxxxx" is not included in request header,  after input token in Authorize.

**Comment by VladislavAntonyuk** [comment]
The same issue. It worked in 1.2.0

**Comment by domaindrivendev** [comment]
Use the following to enable the oauth2 client credentials (aka application) flow:

```csharp
c.AddSecurityDefinition("oauth2", new OAuth2Scheme
{
    Type = "oauth2",
    Flow = "application",
    TokenUrl = "/auth-server/connect/token",
    Scopes = new Dictionary<string, string>
    {
        { "readAccess", "Access read operations" },
        { "writeAccess", "Access write operations" }
    }
});

c.AddSecurityRequirement(new Dictionary<string, IEnumerable<string>>
{
    { "oauth2", new[] { "readAccess", "writeAccess" } }
});
```

**Comment by markpritchett** [comment]
Thanks for your reply @domaindrivendev.

Having pulled the latest code and made the changes above, I now see the following error:

![image](https://user-images.githubusercontent.com/1267889/36915616-4548a8dc-1e49-11e8-8c53-b718011a2b45.png)

Does it work for you?

**Comment by arlan85** [comment]
This don't work for me, after put client id and key on popup. It says below : 
`**auth errorTypeError: NetworkError when attempting to fetch resource.**`

**Comment by deinok** [comment]
Same here

**Comment by coder8keyboard** [comment]
same here. Any documentation for client_credentials flow?

**Comment by VictorioBerra** [comment]
I also need documentation. Im getting the exact same error. Its not requesting the credentials correctly to IdentityServer4.

**Comment by VictorioBerra** [comment]
https://openid.net/specs/openid-connect-core-1_0.html#TokenRequest

> The Client sends the parameters to the Token Endpoint using the HTTP POST method and the Form Serialization, per Section 13.2, as described in Section 4.1.3 of OAuth 2.0 [RFC6749].

This is not what is happening. Its attempting to use basic authentication.

**Comment by alexmartinezm** [comment]
Same here...

**Comment by VictorioBerra** [comment]
Guys his example works. Turns out it is `Flow = "application"` though I second the ask for documentation. Someone can always make a PR.

**Comment by cmreynol** [comment]
Anyone have an updated example of this using OpenApiSecurityScheme?

**Comment by ghost1face** [comment]
> @VictorioBerra Anyone have an updated example of this using OpenApiSecurityScheme?

I'm experimenting with this at the moment, will let you know how it goes:

```csharp
options.AddSecurityDefinition("oauth2", new OpenApiSecurityScheme
{
    Type = SecuritySchemeType.OAuth2,
    Flows = new OpenApiOAuthFlows
    {
         ClientCredentials = new OpenApiOAuthFlow
         {
              TokenUrl = new Uri("/auth/token", UriKind.Relative)
         }
    }
});
```

**Comment by SimonKoegler** [comment]
@ghost1face any updates?

**Comment by ghost1face** [comment]
> @ghost1face any updates?

That actually worked pretty well 👍

**Comment by jstallm** [comment]
> Use the following to enable the oauth2 client credentials (aka application) flow:
> 
> ```cs
> c.AddSecurityDefinition("oauth2", new OAuth2Scheme
> {
>     Type = "oauth2",
>     Flow = "application",
>     TokenUrl = "/auth-server/connect/token",
>     Scopes = new Dictionary<string, string>
>     {
>         { "readAccess", "Access read operations" },
>         { "writeAccess", "Access write operations" }
>     }
> });
> 
> c.AddSecurityRequirement(new Dictionary<string, IEnumerable<string>>
> {
>     { "oauth2", new[] { "readAccess", "writeAccess" } }
> });
> ```

Is this for ASP.NET Core or the full .net framework?

I'm getting that 

> The name OAuth2Scheme does not exist in the current context

**Comment by jstallm** [comment]
> Guys his example works. Turns out it is `Flow = "application"` though I second the ask for documentation. Someone can always make a PR.

Did you validate this yourself using ASP.NET Core framework? If so, where is the class "OAuth2Scheme" defined? Did you define this yourself?

**Comment by SrikanthSamala555** [comment]
OAuth2Scheme  available in Dotnet Core 6.0
